fml/memory/threadsafe_unique_ptr.h

```diff
@@ -0,0 +1,111 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_THREADSAFE_UNIQUE_PTR_H_
+#define FLUTTER_FML_THREADSAFE_UNIQUE_PTR_H_
+
+#include <memory>
+#include <mutex>
+
+#include "flutter/fml/macros.h"
+
+namespace fml {
+
+/// A single-owner smart pointer that allows weak references that can be
+/// accessed on different threads.
+/// This smart-pointer makes the following guarantees:
+///   * There is only ever one threadsafe_unique_ptr per object.
+///   * The object is deleted when ~threadsafe_unique_ptr() is called.
+///   * The object will not be deleted whilst another thread has a lock_ptr.
+///   * The thread that owns the object can access it without a lock.
+/// WARNING: weak_ptr's should only be used to invoke thread-safe methods.
+template <typename T>
+class threadsafe_unique_ptr {
+ private:
+  struct Data {
+    Data(std::unique_ptr<T>&& arg_ptr) : ptr(std::move(arg_ptr)) {}
+    std::unique_ptr<T> ptr;
+    std::mutex mutex;
+  };
+
+ public:
+  threadsafe_unique_ptr()
+      : data_(new Data(std::unique_ptr<T>())), fast_ptr_(nullptr) {}
+
+  threadsafe_unique_ptr(std::unique_ptr<T>&& ptr)
+      : data_(new Data(std::move(ptr))) {
+    fast_ptr_ = data_->ptr.get();
+  }
+
+  threadsafe_unique_ptr(threadsafe_unique_ptr<T>&& ptr)
+      : data_(std::move(ptr.data_)), fast_ptr_(std::move(ptr.fast_ptr_)) {}
+
+  threadsafe_unique_ptr& operator=(threadsafe_unique_ptr&& rvalue) {
+    data_ = std::move(rvalue.data_);
```

Comment:
`data_` may be read and written from multiple threads here and on line 51.

Elsewhere in the engine, if we need a weak_ptr we just use a shared_ptr, maybe with a doc comment stating that this pointer is intended for sharing only as a weak.

I'd prefer to use STL primitives rather than adding new ones to FML - they become vectors for bugs and surprising missing functionality...

Reply:
`threadsafe_unique_ptr` is composed of stl primitives.  The code can be written with just stl primitives directly, and it was previously, but this actually makes the code easier to verify since we have one unit we can test and ensure works instead of carrying around a `std::shared_ptr<std::unique_ptr<>>` and `std::mutex` to represent the same thing.

We cannot use `std::shared_ptr` because it doesn't guarentee which thread the destructor will be called on or when the destructor will be executed.  This has lead to bugs in the past and is key to the mechanism that disables platform channel messages from background isolates while the engine is deallocating.

We can't use `fml::RefPtr` because it is only designed to work across a single thread.

Reply:
outdated